File: Grafos/exemplo-busca-em-largura-lista-de-adjacencias.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
/**
 * A lista de adjacências representa o grafo em espaço O(|V|+|E|)
 * e possibilita busca em largura ou profundidade em tempo O(|V|+|E|)
 */

void bfs(vector<vector<int>>& g, int v){
    const int white = 0;
    const int grey = 1;
    const int black = 2;
    vector<int> color(g.size(),white);
    color[v]=grey;
    queue<int> q;
    q.push(v);
    while(!q.empty()){
        int u = q.front();
        cout << "Nó visitado = " << u << endl;
        q.pop();
        for(size_t j=0;j<g[u].size();j++){
            int w = g[u][j];
            if(color[w]==white){
                color[w] = grey;
                q.push(w);
            }
        }
        color[u] = black;       
    }
}





int main(){
    vector<vector<int>> g;
    g.resize(5);
    g[0] = {1,4};
    g[1] = {2,3};
    g[2] = {3,4};
    g[3] = {3,3};
    g[4] = {0,1};
    bfs(g,4);
    return 0;
}

File: Grafos/_exemplo_busca_em_largura_lista_de_adjacencias.java
package grafos;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class _exemplo_busca_em_largura_lista_de_adjacencias {
	static void dfs(	List<List<Integer>> g, int v){

	    int white = 0; //vértice não processado
	    int grey = 1; //vértice em processamento
	    int black = 2; // vértice processado

	    List<Integer> color = new ArrayList<Integer>(Collections.nCopies(g.size(), white));//cor
	    List<Integer> d = new ArrayList<Integer>(Collections.nCopies(g.size(), Integer.MAX_VALUE));//cor
		List<Integer> pai = new ArrayList<Integer>(Collections.nCopies(g.size(), -1));//informação do pai
		
	    d.set(v, 0);
	    color.set(v, grey);
	    Queue<Integer> q = new LinkedList<Integer>();
	    q.offer(v);
	    while(!q.isEmpty()){
	        int u = q.peek(); // retira o nó da fila
	        q.poll();
	        System.out.println("Nó visitado = "+u);
	        System.out.println("Pai do nó = "+pai.get(u));
	        System.out.println( "Distância da origem "+v+" = "+d.get(u));

	        for(int j=0;j<g.get(u).size();j++){ //percorre a lista de adjacências
	        	 int w = g.get(u).get(j);  //retira o nó adjacente
		            if(color.get(w)==white){  //se o nó ainda não foi visitado, inclue ele na lista
		            	color.set(w, grey); // marca o nó como em processamento
		            	pai.set(w, u); //atualiza o pai
		            	d.set(w, d.get(u)+1); //atualiza a distância
	                q.offer(w); //coloca o nó na fila
	            }
	        }
	        color.set(u, black);        
	    }
	}
	
	public static void main(String[] args) {
		List<List<Integer>> g = new ArrayList<List<Integer>>();
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.get(0).add(1);
		g.get(0).add(4);
		g.get(1).add(2);
		g.get(1).add(3);
		g.get(2).add(3);
		g.get(2).add(4);
		g.get(3).add(3);
		g.get(3).add(3);
		g.get(4).add(0);
		g.get(4).add(1);

		dfs(g, 4);
	}
}
